File: python/infinicore/ops/add_rms_norm.py
import infinicore.tensor as tensor
from infinicore.lib import _infinicore


def add_rms_norm(a, b, weight, epsilon=1e-5, *, out=None, residual=None):
    """
    Fused Add and RMS Normalization.

    Args:
        a: First input tensor
        b: Second input tensor
        weight: Scale weights
        epsilon: Small constant for numerical stability, default is 1e-5
        out: Optional output tuple (y, residual_out) for in-place operation

    Returns:
        Tuple of (normalized_result, add_result): (RMSNorm(a + b) * weight, a + b)
        The add_result can be used as residual for subsequent layers.
    """
    if out is None:
        out = tensor.empty(a.shape, dtype=a.dtype, device=a.device)
    if residual is None:
        residual = tensor.empty(b.shape, dtype=b.dtype, device=b.device)

    _infinicore.add_rms_norm_(
        out._underlying,
        residual._underlying,
        a._underlying,
        b._underlying,
        weight._underlying,
        epsilon,
    )

    return out, residual

File: ops/src/infinicore/ops/add_rms_norm/add_rms_norm.cc
#include "infinicore/ops/add_rms_norm.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(AddRMSNorm);

AddRMSNorm::AddRMSNorm(Tensor y, Tensor residual_out, const Tensor &a, const Tensor &b, const Tensor &weight, float epsilon) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(y, residual_out, a, b, weight);
    INFINICORE_GRAPH_OP_DISPATCH(y->device().getType(), y, residual_out, a, b, weight, epsilon);
}

void AddRMSNorm::execute(Tensor y, Tensor residual_out, const Tensor &a, const Tensor &b, const Tensor &weight, float epsilon) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(AddRMSNorm, y, residual_out, a, b, weight, epsilon);
}

std::pair<Tensor, Tensor> add_rms_norm(const Tensor &a, const Tensor &b, const Tensor &weight, float epsilon) {
    auto y = Tensor::empty(a->shape(), a->dtype(), a->device());
    auto residual_out = Tensor::empty(a->shape(), a->dtype(), a->device());
    add_rms_norm_(y, residual_out, a, b, weight, epsilon);
    return std::make_pair(y, residual_out);
}

void add_rms_norm_(Tensor out, Tensor residual, const Tensor &a, const Tensor &b, const Tensor &weight, float epsilon) {
    AddRMSNorm::execute(out, residual, a, b, weight, epsilon);
}

void add_rms_norm_inplace(Tensor input, Tensor residual, const Tensor &weight, float epsilon) {
    add_rms_norm_(input, residual, input, residual, weight, epsilon);
}

} // namespace infinicore::op
